source/opt/local_single_store_elim_pass.cpp

```diff
@@ -319,6 +120,127 @@ void LocalSingleStoreElimPass::InitExtensions() {
       "SPV_EXT_descriptor_indexing",
   });
 }
+bool LocalSingleStoreElimPass::ProcessVariable(ir::Instruction* var_inst) {
+  vector<ir::Instruction*> users;
+  FindUses(var_inst, &users);
+
+  ir::Instruction* store_inst = FindSingleStoreAndCheckUses(var_inst, users);
+
+  if (store_inst == nullptr) {
+    return false;
+  }
+
+  return RewriteLoads(store_inst, users);
+}
+
+ir::Instruction* LocalSingleStoreElimPass::FindSingleStoreAndCheckUses(
+    ir::Instruction* var_inst, const vector<ir::Instruction*>& users) const {
+  // Make sure there is exactly 1 store.
+  ir::Instruction* store_inst = nullptr;
+
+  // If |var_inst| has an initializer, then that will count as a store.
+  if (var_inst->NumInOperands() > 1) {
+    store_inst = var_inst;
+  }
+
+  for (ir::Instruction* user : users) {
+    switch (user->opcode()) {
+      case SpvOpStore:
+        if (store_inst == nullptr) {
```

Comment:
Don't we need a check here that the variable of interest is the *first* operand of the store? If it isn't, I think this algorithm will generate incorrect code. Yeah, having a variable show up in the second operand here is admittedly rare, but I am pretty sure it is legal, so we should guard against it.  Since it is rare, I haven't thought a lot about the optimal way to deal with it, but I think it would be *safe* to just return nullptr for the function. Since this event is rare, it wouldn't hurt performance all that much.

Reply:
OK